src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java

```diff
@@ -73,26 +84,52 @@ public RedirectView login(final @RequestParam(name = REDIRECT_URL_NAME) Optional
                 && customJwtService.isTokenValid(customJWTCookie.getValue());
 
         if (!isTokenValid) {
-            final Cookie redirectUrlCookie = WebUtil.buildSecureCookie(REDIRECT_URL_NAME,
-                    redirectUrlParam.orElse(configProperties.getDefaultRedirectUrl()));
-            response.addCookie(redirectUrlCookie);
+            final String state = encryptionService.getSHA512SecurePassword(
+                    oneLoginService.generateAndStoreState(response, redirectUrlParam.orElse(configProperties.getDefaultRedirectUrl()))
+            );
+            final String nonce = oneLoginService.generateAndStoreNonce();
 
-            return new RedirectView(oneLoginService.getOneLoginAuthorizeUrl());
+            return new RedirectView(oneLoginService.getOneLoginAuthorizeUrl(state, nonce));
         }
 
         return new RedirectView(redirectUrlParam.orElse(configProperties.getDefaultRedirectUrl()));
     }
 
     @GetMapping("/redirect-after-login")
-    public RedirectView redirectAfterLogin(final @CookieValue(name = REDIRECT_URL_NAME) String redirectUrlCookie,
+    public RedirectView redirectAfterLogin(
+            final @CookieValue(name = STATE_COOKIE) String stateCookie,
             final HttpServletResponse response,
-            final @RequestParam String code) {
-        final OneLoginUserInfoDto userInfo = oneLoginService.getOneLoginUserInfoDto(code);
-        final User user = oneLoginService.createOrGetUserFromInfo(userInfo);
-        addCustomJwtCookie(response, userInfo);
-        return new RedirectView(user.getLoginJourneyState()
-                .getLoginJourneyRedirect(user.getHighestRole().getName())
-                .getRedirectUrl(adminBaseUrl, applicantBaseUrl, techSupportAppBaseUrl, redirectUrlCookie));
+            final @RequestParam String code,
+            final @RequestParam String state
+    ) {
+        final JSONObject tokenResponse = oneLoginService.getOneLoginUserTokenResponse(code);
+        IdTokenDto decodedIdToken = oneLoginService.getDecodedIdToken(tokenResponse);
```

Comment:
This method can be broken down into smaller private ones.  Alternatively create and use a Utils class for doing the extra state and nonce transformations

Reply:
do you mean getDecodedIdToken just, or all the logic I've added to this function?

Reply:
I've split getDecodedIdToken into
```
    private String decodeJWT(final String token) {
        String[] chunks = token.split("\\.");
        Base64.Decoder decoder = Base64.getUrlDecoder();
        return new String(decoder.decode(chunks[1]));
    }

    public IdTokenDto getDecodedIdToken(final JSONObject tokenResponse) {
        final String idToken = tokenResponse.getString("id_token");
        ObjectMapper mapper = new ObjectMapper();
        IdTokenDto decodedIdToken = new IdTokenDto();
        try {
            decodedIdToken = mapper.readValue(decodeJWT(idToken), IdTokenDto.class);
        } catch (JsonProcessingException e) {
            e.printStackTrace();
        }
        return decodedIdToken;
    }
```
But I don't think I can split it any further due to the fact that it's built specifically for idToken